### PR TITLE
Optimistic updates in object renderer

### PR DESCRIPTION
**Pull request by 9at8**
#99635 

Adds optimistic rendering to the object renderer + some other polish stuff.

## Before
after

## After
after

### CODE REVIEW

File: src/vs/workbench/contrib/preferences/browser/settingsWidgets.ts [hunk 1]
```diff
@@ -735,19 +735,15 @@ export class ObjectSettingWidget extends AbstractListSettingWidget<IObjectDataIt
 	}
 
 	protected renderHeader() {
-		if (this.model.items.length > 0) {
```
Comment (9at8) [inline src/vs/workbench/contrib/preferences/browser/settingsWidgets.ts:738]: If we are starting a brand new table, there isn't enough space to render the new row properly. I think we should always render the header.

## Before

image

## After

image

File: src/vs/workbench/contrib/preferences/browser/settingsTree.ts [hunk 2]
```diff
@@ -1080,7 +1087,7 @@ export class SettingObjectRenderer extends AbstractSettingRenderer implements IT
 
 		template.objectWidget.setValue(items, {
 			showAddButton: (
-				isDefined(dataElement.setting.objectAdditionalProperties) ||
+				typeof dataElement.setting.objectAdditionalProperties === 'object' ||
```
Comment (9at8) [inline src/vs/workbench/contrib/preferences/browser/settingsTree.ts:1090]: We don't want to show the add button if `additionalProperties` if `false`

File: src/vs/workbench/contrib/preferences/browser/settingsTree.ts [hunk 3]
```diff
@@ -1035,25 +1035,30 @@ export class SettingObjectRenderer extends AbstractSettingRenderer implements IT
...
 				// Item was updated
-				if (isDefined(e.item) && e.originalItem.key.data === item.key.data) {
+				if (isDefined(e.item) && e.targetIndex === idx) {
 					newValue[e.item.key.data] = e.item.value.data;
+					newItems.push(e.item);
 				}
-				// All remaining items
-				else {
+				// All remaining items, but skip the one that we just updated
+				else if (isUndefinedOrNull(e.item) || e.item.key.data !== item.key.data) {
```
Comment (9at8) [inline src/vs/workbench/contrib/preferences/browser/settingsTree.ts:1047]: Consider we have files associations like this

|Item|Value|
|---|---|
|foo|bar|
|hello|world|

Let's say I change `foo` to `hello`

|Item|Value|
|---|---|
|hello|bar|
|hello|world|

Then the correct table should be:

|Item|Value|
|---|---|
|hello|bar|

**Review by roblourens — APPROVED** [review]
This looks great!

**Pull request merged**